File: shared/bandwidth/composite.py
"""Composite bandwidth estimator - takes minimum of multiple backends."""

from shared.bandwidth.base import BandwidthEstimator


class CompositeBackend(BandwidthEstimator):
    """
    Take the minimum of several backends.
    Ensures we never exceed the most conservative estimate.
    """

    def __init__(self, *backends: BandwidthEstimator):
        self._backends = list(backends)

    def get_target_bps(self) -> float:
        if not self._backends:
            return 10_000  # fallback
        return min(b.get_target_bps() for b in self._backends)

    def on_bytes_sent(self, n: int) -> None:
        for b in self._backends:
            b.on_bytes_sent(n)

    def on_bytes_queued(self, n: int) -> None:
        for b in self._backends:
            b.on_bytes_queued(n)

File: sidecar/internal/bandwidth/composite.go
package bandwidth

// Composite returns the minimum TargetBps across its members. Mirror of the
// Python CompositeBackend: the most conservative estimate wins so we never
// exceed any single backend's read on the link.
type Composite struct {
	members []Estimator
}

func NewComposite(members ...Estimator) *Composite {
	return &Composite{members: members}
}

func (c *Composite) TargetBps() uint64 {
	if len(c.members) == 0 {
		return 10_000
	}
	out := c.members[0].TargetBps()
	for _, m := range c.members[1:] {
		if v := m.TargetBps(); v < out {
			out = v
		}
	}
	return out
}
